### PR TITLE
chore: extract all messaging logic to its own package

### PR DESCRIPTION
**Pull request by smoya**
**Description**

This PR aims to make all messaging logic around the code reusable by any component. A must for WIP as https://github.com/asyncapi/event-gateway/issues/37 and in the future https://github.com/asyncapi/event-gateway/issues/43.

Previously, there was several models for what a Message represents (KafkaMessage, Message...) but this has been unified and extracted to a new package called `message`. There are some renewed concepts introduced as well.
The main changes are:

- New `message` package. Contains models for a Message (including headers or metadata, context, etc), Message validators and Middlewares.
- `kafka.Message` got removed in favour of the new Message in `message` package.
- All message handlers used in the Kafka proxy (there was only one for validating the message) are now in `message` package so they can be reused by any other component, protocol, etc.
- Testing has been improved.  

**Related issue(s)**
https://github.com/asyncapi/event-gateway/issues/37

### CODE REVIEW

File: message/middleware.go [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+// ErrMiddlewareMissingReturn is the error used when a middleware did not return either a message or an error but one is expected.
+var ErrMiddlewareMissingReturn = errors.New("a middleware did not return either a message or an error but one is expected")
+
+// Middleware is a function applied to a Message and returns the same Message after processing it or modifying it.
+// If a middleware does not return a Message, the message should be considered as discarded.
+type Middleware func(m *Message) (*Message, error)
+
+// Chain calls all given middlewares in a pipeline-fashioned process. Order of middlewares is preserved during execution.
+func Chain(middlewares ...Middleware) Middleware {
```
Comment (smoya) [inline message/middleware.go:13]: Was close to being named `Pipeline` but then thought `Middleware` should be called `Pipe` and that's not exactly what it is. Also, `Middleware` is a very well-known term so I won't change that at all.

Any suggestion is welcome.

Comment (magicmatatjahu) [reply message/middleware.go:13]: Chain is good name :) You can also add `Middlewares*` prefix to be more explicit - `MiddlewaresChain`

Comment (smoya) [reply message/middleware.go:13]: Good Point. I'm gonna rename it to `MiddlewaresChain` as suggested 👍

Comment (magicmatatjahu) [reply message/middleware.go]: I wonder if that is a good name for this function. As I see, middlewares are executed sequentially, and usually middlewares (in JS, maybe also in other languages) should have as argument `next` i.e. next middleware - middleware should determine whether to execute next middleware or break the chain.

File: message/message.go [hunk 2]
```diff
@@ -0,0 +1,20 @@
+package message
+
+// Message represents a message flowing through the wire. For example, a Kafka message.
+type Message struct {
+	Context Context  `json:"context"`
+	Key     []byte   `json:"key,omitempty"`
+	Value   []byte   `json:"value,omitempty"`
```
Comment (magicmatatjahu) [inline message/message.go:7]: Is this the same message as in AsyncAPI? I am missing the `payload` field. Unless `payload` is a `value`.

Comment (smoya) [reply message/message.go:7]: No, this is not related to AsyncAPI at all. This is only a representation of message flowing through somewhere, E.g. A Kafka message flowing through a topic partition. But could be also any other protocol.

Comment (smoya) [reply message/message.go:7]: AsyncAPI Models are under https://github.com/asyncapi/event-gateway/tree/master/asyncapi but weren't changed on this PR.

Comment (magicmatatjahu) [reply message/message.go:7]: Ok, thanks! 😄

File: message/middleware.go (hunk 1 above)
Comment (smoya) [reply message/middleware.go]: You are totally right. In fact what I named as Middleware is not Middleware but Handler. Middleware will just get a Handler as input, and Output another Handler.

Comment (smoya) [reply message/middleware.go]: Let's put this PR on-hold as I'm gonna evaluate to use Watermill.io -> https://github.com/asyncapi/event-gateway/issues/43 as It has all we need (routers, handlers and middlewares, publishers, subscribers, etc) and most probably we won't need most of the code in here anymore.

**Comment by smoya** [comment]
Let's close this PR as I'm gonna evaluate to use Watermill.io -> #43 as It has all we need (routers, handlers and middlewares, publishers, subscribers, etc) and most probably we won't need most of the code in here anymore.

Context: https://github.com/asyncapi/event-gateway/pull/49#discussion_r717397777

File: message/middleware.go (hunk 1 above)
Comment (magicmatatjahu) [reply message/middleware.go]: To not block you, we can go with this implementation. Please only add comment about this "problem" ot change name to HandlersChain, or something like that.

Comment (smoya) [reply message/middleware.go]: @magicmatatjahu Middleware has been renamed to Handler, and all related accordingly to what we discussed in here.

Comment (magicmatatjahu) [reply message/middleware.go]: Thanks!

File: kafka/config.go [hunk 3]
```diff
@@ -18,15 +18,17 @@ type ProxyConfig struct {
...
-	MessageHandlers    []MessageHandler
+	MessageHandlers    []message.Handler
 	Debug              bool
 }
 
+// Option represents a functional configuration for the Proxy.
 type Option func(*ProxyConfig) error
 
-func WithMessageHandlers(messageHandlers ...MessageHandler) Option {
+// WithMessageHandlers configures sets a given list of message.Middlware as proxy message handlers.
```
Comment (magicmatatjahu) [inline kafka/config.go:28]:
```suggestion
// WithMessageHandlers configures sets a given list of message.Handler as proxy message handlers.
```

😅

**Review by magicmatatjahu — COMMENTED** [review]
One suggestion 😅

File: kafka/config.go (hunk 3 above)
Comment (smoya) [reply kafka/config.go]: LOL

**Review by magicmatatjahu — APPROVED** [review]

**Pull request merged**